File: algorithms/110.swift
// https://leetcode.com/problems/balanced-binary-tree

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */

class Solution {
    func isBalanced(root: TreeNode?) -> Bool {
        func dfsHeight(node: TreeNode?) -> Int {
            guard let n = node else { return 0 }

            let leftHeight = dfsHeight(n.left)
            if leftHeight == -1 { return -1 }

            let rightHeight = dfsHeight(n.right)
            if rightHeight == -1 { return -1 }

            if abs(leftHeight - rightHeight) > 1 { return -1 }
            return max(leftHeight, rightHeight) + 1
        }
        return dfsHeight(root) != -1
    }
}

File: algorithms/110.py
# https://leetcode.com/problems/balanced-binary-tree


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """

        def dfs_check(node):
            left_height, right_height = 0, 0
            if node is None:
                return (0, True)
            left_height, left_balanced = dfs_check(node.left)
            if not left_balanced:
                return (0, False)
            right_height, right_balanced = dfs_check(node.right)
            if not right_balanced:
                return (0, False)
            return (max(left_height, right_height) + 1, abs(left_height - right_height) <= 1)

        height, balanced = dfs_check(root)
        return balanced
